astroid/interpreter/_import/util.py

```diff
@@ -2,15 +2,64 @@
 # For details: https://github.com/PyCQA/astroid/blob/main/LICENSE
 # Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
 
-try:
-    import pkg_resources
-except ImportError:
-    pkg_resources = None  # type: ignore[assignment]
-
-
-def is_namespace(modname):
-    return (
-        pkg_resources is not None
-        and hasattr(pkg_resources, "_namespace_packages")
-        and modname in pkg_resources._namespace_packages
-    )
+from __future__ import annotations
+
+import os
+import pathlib
+from functools import lru_cache
+from importlib.util import _find_spec_from_path
+
+
+def _is_setuptools_namespace(location: str | pathlib.Path) -> bool:
+    try:
+        with open(os.path.join(location, "__init__.py"), "rb") as stream:
+            data = stream.read(4096)
+    except OSError:
+        return False
+    else:
+        extend_path = b"pkgutil" in data and b"extend_path" in data
+        declare_namespace = (
+            b"pkg_resources" in data and b"declare_namespace(__name__)" in data
+        )
+        return extend_path or declare_namespace
+
+
+@lru_cache(maxsize=4096)
+def is_namespace(modname: str) -> bool:
+    found_spec = None
+
+    # find_spec() attempts to import parent packages when given dotted paths.
+    # That's unacceptable here, so we fallback to _find_spec_from_path(), which does
+    # not, but requires instead that each single parent ('astroid', 'nodes', etc.)
+    # be specced from left to right.
+    working_modname = ""
+    last_parent = None
+    for component in modname.split("."):
+        if working_modname:
```

Comment:
One last comment just because I wonder if you know the answer:
What would be the performance impact of only assigning `component` here and adding the `.` on L47-48.

It feels a bit wasteful to have an `if` check that is only really necessary for the first iteration, but on the other hand an additional assignment would also be costly.
Just something I thought about and to which I really don't have the answer.

Anyway, such optimisations are really not what we're after (right now). Just a thought experiment 😄